docs/04-refactor-state-management.md

```diff
@@ -91,126 +90,125 @@ public void ResetOrder()
-``` 
-
-This is why when we defined `ConfirmConfigurePizzaDialog` we put in a call to `StateHasChanged`. `StateHasChanged` is the way to trigger a manual re-render.
-
-For now the easiest way to think about it is... *if I am passing a delegate to another component, and I want it to update my UI, reach for StateHasChanged*. This pattern is fairly common so it's important to know about. We're looking to improve the situation further in the future.
-
-In our case, we've still got a problem because the methods we're passing around are defined on the `OrderState`. So we need a way for `OrderState` to trigger an update on any component that depends upon it. 
 
--------
+In the this example the event handler delegate is `TestComponent2.Clicked` and the `Delegate.Target` is the instance of `TestComponent` - even though it's `CoolButton` that actually defines the event. This means that `TestComponent2` will be rerendered when the button is clicked. This makes sense because if `TestComponent2` didn't get rerendered, you couldn't update the count.
 
-Next, let's make it possible for `OrderState` to produce its own state-change notifications.
+Let's see a third example to show how this falls apart. There are cases where `Delegate.Target` isn't a component at all, and so nothing will rerender. Let's see that example again, but with an *AppState* object:
 
-Let's define a .NET event on `OrderState`
 ```C#
-public event EventHandler StateChanged;
-
-private void StateHasChanged()
+public class TestState
 {
-    StateChanged?.Invoke(this, EventArgs.Empty);
+    public int Count { get; private set; }
+
+    public void Clicked()
+    {
+        Count++;
+    }
 }
 ```
 
-Now we can add a call to `StateHasChanged` to methods on `OrderState` that need it (`CancelConfigurePizzaDialog` and `ConfirmConfigurePizzaDialog`, and `RemoveConfiguredPizza`).
+```html
+@* CoolButton.razor *@
+<button onclick="Clicked">Clicking this will be cool!</button>
+@functions {
+    [Parameter] Action Clicked { get; set; }
+}
 
-To hook up the other side, we should subscribe to the change notifications on `Index`
+@* TestComponent3.razor *@
+@inject TestState State
+<CoolButton Clicked="@State.Clicked" />
+<p>Clicked @State.Count times!</p>
 ```
-@implements IDisposable
 
-...
+In this third example the event handler delegate is `TestState.Clicked` and the so `Delegate.Target` is `TestState` - **not a component**. When the button is clicked, no component gets the event notification, and so nothing will rerender.
 
-@functions {
-    ...
+This is the problem that `EventCallback` was created to solve. By changing the parameter on `CoolButton` from `Action` to `EventCallback` you fix the event dispatching behavior. This works because `EventCallback` is known to the compiler, when you create an `EventCallback` from a delegate that doesn't have its `Target` set to a component, then the compiler will pass the curent component to recieve the event.
```

Comment:
I'm open to suggestions about how to improve this. I didn't feel like I had a good way to explain `EventCallback` without the 3 examples here. Does this seem good? Other ideas?

Reply:
Typo: `curent`

Rather than having this as a long written explanation, we could consider making this one of the demos we give. That would give people a chance to ask questions and wouldn't risk them thinking they were meant to add any of this code to their project.

Alternatively it could be moved to the end of the document as an "appendix" and referenced from here, where at this point in the flow we just say basically "you should always use `EventCallback` for callback parameters, because it makes event dispatching do what you want - see appendix 1 for technical details".

Reply:
I added a terse explanation and moved this to the speakers notes since the examples aren't totally connected to the code in the workshop.